modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  The :mod:`Initialization` module provides functions which enable users to
+  move-initialize and deinitialize values.
+
+  The functionality provided by this module can be used to implement
+  collections in a manner similar to those implemented by the Chapel standard
+  modules (such as :mod:`List` or :mod:`Map`).
+*/
+module Initialization {
+
+  private inline proc _move(ref dst: ?t, const ref src: t) {
+    __primitive("=", dst, src);
+  }
+
+  private inline proc _destroy(ref r) {
+    chpl__autoDestroy(r);
+  }
+
+  private inline proc _needsAutoDestroy(type t) param: bool {
+    return __primitive("needs auto destroy", t);
+  }
+
+  /*
+    Check to see if a given type supports deinitialization.
+
+    :arg t: A type to check for a deinitializer
+    :type t: `type`
+
+    :return: ``true`` if ``t`` supports deinitialization
+    :rtype: param bool
+  */
+  proc needsDeinit(type t) param {
+    return _needsAutoDestroy(t);
+  }
+
+  /*
+    Explicitly deinitialize a reference. The memory referred to by ``arg``
+    should be considered uninitialized after a call to this function.
+
+    .. warning::
+
+      Chapel is not aware of a call to ``explicitDeinit()``, which means
+      that the deinitializer of ``arg`` may run again once it goes out of
+      scope.
+
+    :arg: A value to deinitialize
+  */
+  proc explicitDeinit(ref arg: ?t) {
+    if needsDeinit(t) then
+      _destroy(arg);
+  }
+
+  /*
+    Move-initialize a reference from a value. No assignment is performed
+    when moving ``rhs`` into the memory referred to by ``lhs``.
+
+    .. warning::
+
+      If the value at ``lhs`` is already initialized memory, it will be
+      overwritten by the contents of ``rhs`` without being deinitialized.
+      Call :proc:`explicitDeinit()` to deinitialize ``lhs`` first if
+      necessary.
+
+    :arg lhs: A reference to move-initialize
+    :type lhs: `?t`
+
+    :arg rhs: A 
+    :type rhs: `t`
+  */
+  proc moveInitialize(ref lhs: ?t, pragma "no auto destroy" in rhs: t) {
+    if lhs.type != nothing then
+      _move(lhs, rhs);
+  }
+
+  /*
+    Move the memory referred to by a reference into a new value.
+
+    .. warning::
+
+      The memory referred to by ``arg`` should be considered uninitialized
+      after a call to this function.
+
+    :arg arg: A reference to consume
+    :type arg: `?t`
+
+    :return: A reference moved into a new value
+    :rtype: `t`
+  */
+  proc moveToValue(const ref arg: ?t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var result: t;
+      _move(result, arg);
+      return result;
+    }
+  }
+
+  /*
+    Swap the values referred to by two references. No assignment takes
+    place when swapping ``lhs`` and ``rhs``.
```

Comment:
... referred to by ``lhs`` and ``rhs``. By contrast with the operator `<=>`,
`moveSwap` does not perform assignments and initializations.

Reply:
I think the comment needs to make it clear that this function also does not call `<=>`.